pkg/agent/nodeportlocal/portcache/port_table.go

```diff
@@ -85,72 +112,104 @@ func (pt *PortTable) GetDataForPodIP(ip string) []NodePortData {
 	return allData
 }
 
-func (pt *PortTable) GetEntryByPodIPPort(ip string, port int) *NodePortData {
+func (pt *PortTable) GetEntryByPodIPPortProtocol(ip string, port int, protocol string) *NodePortData {
 	pt.tableLock.RLock()
 	defer pt.tableLock.RUnlock()
-	return pt.getEntryByPodIPPort(ip, port)
+	return pt.getEntryByPodIPPortProtocol(ip, port, protocol)
 }
 
-func (pt *PortTable) getEntryByPodIPPort(ip string, port int) *NodePortData {
+func (pt *PortTable) getEntryByPodIPPortProtocol(ip string, port int, protocol string) *NodePortData {
 	for _, data := range pt.Table {
-		if data.PodIP == ip && data.PodPort == port {
+		if data.PodIP == ip && data.PodPort == port && data.HasProtocol(protocol) {
 			return &data
 		}
 	}
 	return nil
 }
 
-func (pt *PortTable) getFreePort(podIP string, podPort int) (int, Closeable, error) {
+func (pt *PortTable) isNodePortAvilableForPodIPProtocol(ip string, nodeport int, podPort int, protocol string) bool {
+	val, ok := pt.Table[nodeport]
+	// A given nodeport, is marked unavailable if an entry for it already exists in the PortTable Table and
+	// the nodeport suggested is being used by another Pod or,
+	// the nodeport suggested for a Pod has a different podPort binding or,
+	// the nodeport suggested for a Pod:podPort has the protocol already configured.
+	if ok && (val.PodIP != ip ||
+		(val.PodIP == ip && val.PodPort != podPort) ||
+		(val.PodIP == ip && val.PodPort == podPort && val.HasProtocol(protocol))) {
+		return false
+	}
+	return true
+}
+
+func (pt *PortTable) getFreePort(podIP string, podPort int, protocol string) (int, Closeable, error) {
 	for i := pt.StartPort; i <= pt.EndPort; i++ {
-		if _, ok := pt.Table[i]; !ok {
-			socket, err := pt.LocalPortOpener.OpenLocalPort(i)
-			if err != nil {
-				continue
-			}
-			return i, socket, nil
+		// This ensures that the NodePort `i` is not taken by any other Pods,
+		// irrespective of the protocol.
+		if !pt.isNodePortAvilableForPodIPProtocol(podIP, i, podPort, protocol) {
+			continue
+		}
+		socket, err := pt.LocalPortOpener.OpenLocalPort(i, protocol)
+		if err != nil {
+			continue
 		}
+		return i, socket, nil
 	}
 	return 0, nil, fmt.Errorf("no free port found")
 }
 
-func (pt *PortTable) AddRule(podIP string, podPort int) (int, error) {
+func (pt *PortTable) AddRule(podIP string, podPort int, protocol string) (int, error) {
 	pt.tableLock.Lock()
 	defer pt.tableLock.Unlock()
-	nodePort, socket, err := pt.getFreePort(podIP, podPort)
+	nodePort, socket, err := pt.getFreePort(podIP, podPort, protocol)
```

Comment:
I understand the code, but I think it's time to add a reverse table to map from "podIP + podPort" to `NodePortData` instance. It will make the code more readable and we will no longer need to iterate through the port table for several operations.

Reply:
You mean in addition to the `nodeport` to `NodePortData` map, right?

Reply:
yes that's what I mean

(sorry for the delay, I came back from PTO today)